File: python/agentpin/revocation.py
"""Revocation document handling for AgentPin."""

from datetime import datetime, timezone

from .types import AgentPinError, ErrorCode


def build_revocation_document(entity: str) -> dict:
    """Build an empty revocation document."""
    return {
        "agentpin_version": "0.1",
        "entity": entity,
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "revoked_credentials": [],
        "revoked_agents": [],
        "revoked_keys": [],
    }


def add_revoked_credential(doc: dict, jti: str, reason: str) -> None:
    """Add a revoked credential to the document."""
    doc["revoked_credentials"].append({
        "jti": jti,
        "revoked_at": datetime.now(timezone.utc).isoformat(),
        "reason": reason,
    })
    doc["updated_at"] = datetime.now(timezone.utc).isoformat()


def add_revoked_agent(doc: dict, agent_id: str, reason: str) -> None:
    """Add a revoked agent to the document."""
    doc["revoked_agents"].append({
        "agent_id": agent_id,
        "revoked_at": datetime.now(timezone.utc).isoformat(),
        "reason": reason,
    })
    doc["updated_at"] = datetime.now(timezone.utc).isoformat()


def add_revoked_key(doc: dict, kid: str, reason: str) -> None:
    """Add a revoked key to the document."""
    doc["revoked_keys"].append({
        "kid": kid,
        "revoked_at": datetime.now(timezone.utc).isoformat(),
        "reason": reason,
    })
    doc["updated_at"] = datetime.now(timezone.utc).isoformat()


def check_revocation(doc: dict, jti: str, agent_id: str, kid: str) -> None:
    """Check if a credential, agent, or key is revoked.

    Raises:
        AgentPinError: if revoked
    """
    for rc in doc.get("revoked_credentials", []):
        if rc["jti"] == jti:
            raise AgentPinError(ErrorCode.CREDENTIAL_REVOKED, f"Credential {jti} revoked: {rc['reason']}")

    for ra in doc.get("revoked_agents", []):
        if ra["agent_id"] == agent_id:
            raise AgentPinError(ErrorCode.AGENT_INACTIVE, f"Agent {agent_id} revoked: {ra['reason']}")

    for rk in doc.get("revoked_keys", []):
        if rk["kid"] == kid:
            raise AgentPinError(ErrorCode.KEY_REVOKED, f"Key {kid} revoked: {rk['reason']}")


def fetch_revocation_document(url: str) -> dict:
    """Fetch a revocation document from a URL."""
    import requests

    resp = requests.get(url, headers={"Accept": "application/json"}, timeout=10)

    if not resp.ok:
        raise AgentPinError(ErrorCode.DISCOVERY_FETCH_FAILED, f"HTTP {resp.status_code} fetching {url}")

    return resp.json()

File: javascript/src/revocation.js
/**
 * Revocation document handling for AgentPin.
 */

import { AgentPinError, ErrorCode } from './types.js';

/**
 * Build an empty revocation document.
 * @param {string} entity
 * @returns {object}
 */
export function buildRevocationDocument(entity) {
    return {
        agentpin_version: '0.1',
        entity,
        updated_at: new Date().toISOString(),
        revoked_credentials: [],
        revoked_agents: [],
        revoked_keys: [],
    };
}

/**
 * Add a revoked credential to the document.
 * @param {object} doc
 * @param {string} jti
 * @param {string} reason - One of RevocationReason values
 */
export function addRevokedCredential(doc, jti, reason) {
    doc.revoked_credentials.push({
        jti,
        revoked_at: new Date().toISOString(),
        reason,
    });
    doc.updated_at = new Date().toISOString();
}

/**
 * Add a revoked agent to the document.
 * @param {object} doc
 * @param {string} agentId
 * @param {string} reason
 */
export function addRevokedAgent(doc, agentId, reason) {
    doc.revoked_agents.push({
        agent_id: agentId,
        revoked_at: new Date().toISOString(),
        reason,
    });
    doc.updated_at = new Date().toISOString();
}

/**
 * Add a revoked key to the document.
 * @param {object} doc
 * @param {string} kid
 * @param {string} reason
 */
export function addRevokedKey(doc, kid, reason) {
    doc.revoked_keys.push({
        kid,
        revoked_at: new Date().toISOString(),
        reason,
    });
    doc.updated_at = new Date().toISOString();
}

/**
 * Check if a credential, agent, or key is revoked.
 * @param {object} doc - Revocation document
 * @param {string} jti
 * @param {string} agentId
 * @param {string} kid
 * @throws {AgentPinError} if revoked
 */
export function checkRevocation(doc, jti, agentId, kid) {
    const rc = doc.revoked_credentials.find(r => r.jti === jti);
    if (rc) {
        throw new AgentPinError(
            ErrorCode.CREDENTIAL_REVOKED,
            `Credential ${jti} revoked: ${rc.reason}`
        );
    }

    const ra = doc.revoked_agents.find(r => r.agent_id === agentId);
    if (ra) {
        throw new AgentPinError(
            ErrorCode.AGENT_INACTIVE,
            `Agent ${agentId} revoked: ${ra.reason}`
        );
    }

    const rk = doc.revoked_keys.find(r => r.kid === kid);
    if (rk) {
        throw new AgentPinError(
            ErrorCode.KEY_REVOKED,
            `Key ${kid} revoked: ${rk.reason}`
        );
    }
}

/**
 * Fetch a revocation document from a URL.
 * @param {string} url
 * @returns {Promise<object>}
 */
export async function fetchRevocationDocument(url) {
    const response = await fetch(url, {
        headers: { 'Accept': 'application/json' },
    });

    if (!response.ok) {
        throw new AgentPinError(
            ErrorCode.DISCOVERY_FETCH_FAILED,
            `HTTP ${response.status} fetching ${url}`
        );
    }

    return response.json();
}
